File: ts/src/onto/parse/bpmn.ts
/**
 * BPMN 2.0 XML -> 可检索、可追溯的流程证据。
 *
 * 前线材料里的 BPMN 文件通常比流程截图更有价值：节点 id、分支条件和连线关系都
 * 是确定性结构，不该再交给模型从图片里猜。本解析器只读取建模所需的流程、活动、
 * 事件、网关和 sequenceFlow；未知扩展保留在原文件中，但不会伪装成已解析内容。
 *
 * ── 为什么这里的每个字段都不能"差不多" ──────────────────────────
 *
 * `structured.nodes` / `structured.sequenceFlows` 是下游流程可视化的**唯一输入**。
 * `sourceRef` / `targetRef` 反了，画出来的箭头就是反的；`type` 从 exclusiveGateway
 * 错成 parallelGateway，读图的人会以为两条分支同时发生。所以：
 *
 *   · 边的方向照抄属性，**不做任何"看起来更合理"的纠正**；
 *   · 网关/事件/任务的 `type` 保留原始标签名，`category` 只是分桶；
 *   · 指不到已解析节点的 ref **发 finding 而不是丢弃**——丢弃会让图看起来完整，
 *     而"看起来完整但少了一条边"是最难被发现的错。
 *
 * XML 选型见 `doc/xmlet.ts` 的文件头（手写零依赖的 ElementTree 等价物）。
 */

import { readFile } from "node:fs/promises";
import { basename } from "node:path";

import { pyRepr } from "../../kernel/errors.js";
import type { ParsedDoc } from "./base.js";
import { Parser, makeChunk, makeFinding, makeParsedDoc } from "./base.js";
import { pyNormalizeSpaces, pyStrip } from "./doc/pycompat.js";
import type { XElement } from "./doc/xmlet.js";
import {
  XmlParseError,
  childrenNamed,
  fromString,
  iterElements,
  itertext,
  localName,
  namespaceOf,
} from "./doc/xmlet.js";
import { hasDtdMarker } from "./doc/ziplite.js";

const MAX_XML_BYTES = 20 * 1024 * 1024;

const TASK_TYPES = new Set([
  "task", "userTask", "serviceTask", "manualTask", "scriptTask",
  "businessRuleTask", "sendTask", "receiveTask", "callActivity", "subProcess",
]);
const EVENT_TYPES = new Set([
  "startEvent", "endEvent", "intermediateCatchEvent", "intermediateThrowEvent",
  "boundaryEvent",
]);
const GATEWAY_TYPES = new Set([
  "exclusiveGateway", "inclusiveGateway", "parallelGateway", "complexGateway",
  "eventBasedGateway",
]);

export interface BpmnLane {
  id: string;
  name: string;
}

export interface BpmnEventDefinition {
  type: string;
  attributes: Record<string, string>;
  text: string;
}

export interface BpmnNode {
  id: string;
  name: string;
  /** 原始标签名（`userTask` / `exclusiveGateway` / …）。画图靠它区分形状。 */
  type: string;
  /** task | event | gateway */
  category: string;
  process_id: string;
  incoming: string[];
  outgoing: string[];
  documentation: string;
  event_definitions: BpmnEventDefinition[];
  lanes: BpmnLane[];
  attributes: Record<string, string>;
}

export interface BpmnFlow {
  id: string;
  name: string;
  type: string;
  process_id: string;
  sourceRef: string;
  targetRef: string;
  condition: string;
  documentation: string;
  attributes: Record<string, string>;
}

export interface BpmnProcess {
  id: string;
  name: string;
  /** 属性缺失时是 `null`（Python 的 `None`），**不是 false** —— "没写"和"写了
   *  false"在导出 BPMN 时是两种不同的行为。 */
  isExecutable: boolean | null;
  documentation: string;
  nodes: BpmnNode[];
  sequenceFlows: BpmnFlow[];
}

/** 解析 `.bpmn` 和复合后缀 `.bpmn20.xml`。 */
export class BpmnParser extends Parser {
  override readonly kind = "bpmn";
  override readonly extensions = [".bpmn", ".bpmn20.xml"];

  /** 不把普通 `.xml` 抢过来；基础的"取扩展名"认不出复合后缀。 */
  override accepts(path: string): boolean {
    const name = basename(path).toLowerCase();
    return name.endsWith(".bpmn") || name.endsWith(".bpmn20.xml");
  }

  override async parse(path: string, opts: { fileId: string }): Promise<ParsedDoc> {
    const fileId = opts.fileId;
    const fileName = basename(path);
    const doc = makeParsedDoc({ fileId, fileName, kind: this.kind });

    let data: Buffer;
    try {
      data = await readFile(path);
    } catch (e) {
      doc.findings.push(makeFinding(
        "parse_failed", `无法读取 BPMN 文件：${errText(e)}`, {}, "warn"));
      return doc;
    }

    if (data.length > MAX_XML_BYTES) {
      doc.findings.push(makeFinding(
        "file_too_large",
        `BPMN XML 大于 ${Math.floor(Math.floor(MAX_XML_BYTES / 1024) / 1024)} MiB，已拒绝解析`,
        { kind: "xml", pointer: "/" }, "warn"));
      return doc;
    }

    // 解析器不会主动取外部 DTD，但内部实体仍可能放大。BPMN 2.0 实例不需要 DTD，
    // 直接拒绝比尝试区分合法/恶意声明更稳妥。
    if (hasDtdMarker(data)) {
      doc.findings.push(makeFinding(
        "unsafe_xml", "BPMN 含 DTD/ENTITY 声明，已为避免实体展开攻击而拒绝解析",
        { kind: "xml", pointer: "/" }, "warn"));
      return doc;
    }

    let root: XElement;
    try {
      root = fromString(data);
    } catch (e) {
      const locator: Record<string, unknown> = { kind: "xml", pointer: "/" };
      if (e instanceof XmlParseError) {
        locator["line"] = e.position[0];
        locator["column"] = e.position[1];
      }
      doc.findings.push(makeFinding(
        "parse_failed", `BPMN XML 语法错误：${errText(e)}`, locator, "warn"));
      return doc;
    }

    const rootName = localName(root.tag);
    doc.meta["namespace"] = namespaceOf(root.tag);
    doc.meta["definitions_id"] = root.attrib["id"] ?? "";
    doc.meta["target_namespace"] = root.attrib["targetNamespace"] ?? "";
    if (rootName !== "definitions") {
      doc.findings.push(makeFinding(
        "unexpected_root", `BPMN 根节点应为 definitions，实际为 ${pyRepr(rootName)}`,
        { kind: "xml", pointer: `/${rootName}` }, "warn"));
    }

    const processes: BpmnProcess[] = [];
    const allNodes: BpmnNode[] = [];
    const allFlows: BpmnFlow[] = [];
    let order = 0;

    const processElements = [...iterElements(root)]
      .filter((e) => localName(e.tag) === "process");
    for (const [processIndex, process] of processElements.entries()) {
      const processNo = processIndex + 1;
      const processId = process.attrib["id"] || `process-${processNo}`;
      const processName = process.attrib["name"] || processId;
      if (!("id" in process.attrib)) {
        doc.findings.push(makeFinding(
          "missing_id", `第 ${processNo} 个 process 没有 id，临时使用 ${processId}`,
          locatorOf(processId, "process", processId), "warn"));
      }

      const lanes = laneMembers(process);
      const nodes: BpmnNode[] = [];
      const flows: BpmnFlow[] = [];
      let anonymous = 0;
      for (const element of iterElements(process)) {
        const elementType = localName(element.tag);
        const category = categoryOf(elementType);
        if (category === null) continue;
        let elementId = element.attrib["id"] ?? "";
        if (!elementId) {
          anonymous += 1;
          elementId = `anonymous-${category}-${anonymous}`;
          doc.findings.push(makeFinding(
            "missing_id",
            `流程 ${processId} 的 ${elementType} 没有 id，临时使用 ${elementId}`,
            locatorOf(processId, elementType, elementId), "warn"));
        }
        nodes.push({
          id: elementId,
          name: element.attrib["name"] || elementId,
          type: elementType,
          category,
          process_id: processId,
          incoming: childTexts(element, "incoming"),
          outgoing: childTexts(element, "outgoing"),
          documentation: childText(element, "documentation"),
          event_definitions: eventDefinitions(element),
          lanes: [...(lanes.get(elementId) ?? [])],
          attributes: strippedAttributes(element, ["id", "name"]),
        });
      }

      const nodeById = new Map<string, BpmnNode>();
      for (const node of nodes) nodeById.set(node.id, node);

      for (const element of iterElements(process)) {
        if (localName(element.tag) !== "sequenceFlow") continue;
        const flowId = element.attrib["id"] || `flow-${flows.length + 1}`;
        if (!("id" in element.attrib)) {
          doc.findings.push(makeFinding(
            "missing_id",
            `流程 ${processId} 的 sequenceFlow 没有 id，临时使用 ${flowId}`,
            locatorOf(processId, "sequenceFlow", flowId), "warn"));
        }
        const source = element.attrib["sourceRef"] ?? "";
        const target = element.attrib["targetRef"] ?? "";
        flows.push({
          id: flowId,
          name: element.attrib["name"] || flowId,
          type: "sequenceFlow",
          process_id: processId,
          sourceRef: source,
          targetRef: target,
          condition: childText(element, "conditionExpression"),
          documentation: childText(element, "documentation"),
          attributes: strippedAttributes(
            element, ["id", "name", "sourceRef", "targetRef"]),
        });
        for (const [refName, ref] of [["sourceRef", source], ["targetRef", target]] as const) {
          if (!ref || !nodeById.has(ref)) {
            doc.findings.push(makeFinding(
              "dangling_reference",
              `sequenceFlow ${flowId} 的 ${refName}=${pyRepr(ref)} 未指向已解析节点`,
              locatorOf(processId, "sequenceFlow", flowId), "warn"));
          }
        }
      }

      const processData: BpmnProcess = {
        id: processId,
        name: processName,
        isExecutable: boolAttr(process.attrib["isExecutable"]),
        documentation: childText(process, "documentation"),
        nodes,
        sequenceFlows: flows,
      };
      processes.push(processData);
      allNodes.push(...nodes);
      allFlows.push(...flows);

      let processRender = `BPMN 流程 ${processName}（id=${processId}）：`
        + `${nodes.length} 个节点，${flows.length} 条 sequenceFlow`;
      if (processData.documentation) processRender += `。${processData.documentation}`;
      doc.chunks.push(makeChunk({
        docId: `bpmn:${processId}`, fileId, fileName,
        locator: locatorOf(processId, "process", processId),
        render: processRender, raw: processData, order, tags: ["bpmn", "process"],
      }));
      order += 1;

      for (const node of nodes) {
        const laneText = node.lanes.length > 0
          ? "；泳道=" + node.lanes.map((lane) => lane.name).join("、") : "";
        let relationText = "";
        if (node.incoming.length > 0) relationText += "；incoming=" + node.incoming.join("、");
        if (node.outgoing.length > 0) relationText += "；outgoing=" + node.outgoing.join("、");
        const docText = node.documentation ? `；说明=${node.documentation}` : "";
        const definitionText = node.event_definitions.length > 0
          ? "；事件定义=" + node.event_definitions.map((item) => item.type).join("、") : "";
        doc.chunks.push(makeChunk({
          docId: `bpmn:${processId}:${node.id}`, fileId, fileName,
          locator: locatorOf(processId, node.type, node.id),
          render: `BPMN ${node.category} ${node.name}（${node.type}，id=${node.id}）`
            + `；流程=${processName}${laneText}${relationText}${definitionText}${docText}`,
          raw: node, order, tags: ["bpmn", node.category, node.type],
        }));
        order += 1;
      }

      for (const flow of flows) {
        const sourceName = nodeById.get(flow.sourceRef)?.name ?? flow.sourceRef;
        const targetName = nodeById.get(flow.targetRef)?.name ?? flow.targetRef;
        const conditionText = flow.condition ? `；条件=${flow.condition}` : "";
        const labelText = flow.name !== flow.id ? `；名称=${flow.name}` : "";
        doc.chunks.push(makeChunk({
          docId: `bpmn:${processId}:${flow.id}`, fileId, fileName,
          locator: locatorOf(processId, "sequenceFlow", flow.id),
          render: `BPMN sequenceFlow ${flow.id}：${sourceName}`
            + `（${flow.sourceRef}） → ${targetName}（${flow.targetRef}）`
            + `${labelText}${conditionText}`,
          raw: flow, order,
          tags: flow.condition
            ? ["bpmn", "sequenceFlow", "relation", "rule"]
            : ["bpmn", "sequenceFlow", "relation"],
        }));
        order += 1;
      }
    }

    doc.structured = {
      processes,
      nodes: allNodes,
      sequenceFlows: allFlows,
    };
    if (processes.length === 0) {
      doc.findings.push(makeFinding(
        "no_processes", "BPMN 文件里没有 process", { kind: "xml", pointer: "/" }, "warn"));
    }
    return doc;
  }
}

function categoryOf(elementType: string): string | null {
  if (TASK_TYPES.has(elementType) || elementType.endsWith("Task")) return "task";
  if (EVENT_TYPES.has(elementType) || elementType.endsWith("Event")) return "event";
  if (GATEWAY_TYPES.has(elementType) || elementType.endsWith("Gateway")) return "gateway";
  return null;
}

/** `" ".join("".join(child.itertext()).split())` —— 换行和缩进在 XML 里没有语义。 */
function childText(element: XElement, childName: string): string {
  const child = childrenNamed(element, childName)[0];
  return child === undefined ? "" : pyNormalizeSpaces(itertext(child));
}

function childTexts(element: XElement, childName: string): string[] {
  return childrenNamed(element, childName)
    .map((child) => pyNormalizeSpaces(itertext(child)))
    .filter((text) => text);
}

function eventDefinitions(element: XElement): BpmnEventDefinition[] {
  const out: BpmnEventDefinition[] = [];
  for (const child of element.children) {
    const childType = localName(child.tag);
    if (!childType.endsWith("EventDefinition")) continue;
    out.push({
      type: childType,
      attributes: strippedAttributes(child, []),
      text: pyNormalizeSpaces(itertext(child)),
    });
  }
  return out;
}

/** 属性名去掉命名空间（`camunda:async` → `async`），并剔掉已经单列的那几个。 */
function strippedAttributes(element: XElement, skip: readonly string[]): Record<string, string> {
  const skipSet = new Set(skip);
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(element.attrib)) {
    const local = localName(key);
    if (!skipSet.has(local)) out[local] = value;
  }
  return out;
}

function laneMembers(process: XElement): Map<string, BpmnLane[]> {
  const out = new Map<string, BpmnLane[]>();
  for (const lane of iterElements(process)) {
    if (localName(lane.tag) !== "lane") continue;
    const laneId = lane.attrib["id"] ?? "";
    const laneName = lane.attrib["name"] || laneId;
    for (const child of lane.children) {
      if (localName(child.tag) !== "flowNodeRef") continue;
      const ref = pyNormalizeSpaces(itertext(child));
      if (!ref) continue;
      const bucket = out.get(ref);
      if (bucket === undefined) out.set(ref, [{ id: laneId, name: laneName }]);
      else bucket.push({ id: laneId, name: laneName });
    }
  }
  return out;
}

function locatorOf(
  processId: string, elementType: string, elementId: string,
): Record<string, unknown> {
  let pointer = `/definitions/process[@id="${processId}"]`;
  if (elementType !== "process") pointer += `//${elementType}[@id="${elementId}"]`;
  return {
    kind: "xml",
    pointer,
    process: processId,
    element: elementType,
    id: elementId,
  };
}

/** 属性没写是 `null`，写了才看内容。`"TRUE"` 认，`"yes"` 不认 —— 照 Python 的判据。 */
function boolAttr(value: string | undefined): boolean | null {
  if (value === undefined) return null;
  const t = pyStrip(value).toLowerCase();
  return t === "true" || t === "1";
}

function errText(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

File: src/ontocopilot/onto/parse/bpmn.py
"""BPMN 2.0 XML -> 可检索、可追溯的流程证据。

前线材料里的 BPMN 文件通常比流程截图更有价值：节点 id、分支条件和连线关系都
是确定性结构，不该再交给模型从图片里猜。本解析器只读取建模所需的流程、活动、
事件、网关和 sequenceFlow；未知扩展保留在原文件中，但不会伪装成已解析内容。
"""

from __future__ import annotations

from pathlib import Path
from typing import Any
from xml.etree import ElementTree as ET

from .base import Finding, ParsedDoc, Parser, make_chunk

_MAX_XML_BYTES = 20 * 1024 * 1024

_TASK_TYPES = frozenset({
    "task", "userTask", "serviceTask", "manualTask", "scriptTask",
    "businessRuleTask", "sendTask", "receiveTask", "callActivity", "subProcess",
})
_EVENT_TYPES = frozenset({
    "startEvent", "endEvent", "intermediateCatchEvent", "intermediateThrowEvent",
    "boundaryEvent",
})
_GATEWAY_TYPES = frozenset({
    "exclusiveGateway", "inclusiveGateway", "parallelGateway", "complexGateway",
    "eventBasedGateway",
})


class BpmnParser(Parser):
    """解析 ``.bpmn`` 和复合后缀 ``.bpmn20.xml``。"""

    kind = "bpmn"
    extensions = (".bpmn", ".bpmn20.xml")

    def accepts(self, path: Path) -> bool:
        """不把普通 ``.xml`` 抢过来；基础 ``Path.suffix`` 不认识复合后缀。"""
        name = path.name.lower()
        return name.endswith((".bpmn", ".bpmn20.xml"))

    def parse(self, path: Path, *, file_id: str) -> ParsedDoc:
        doc = ParsedDoc(file_id=file_id, file_name=path.name, kind=self.kind)
        try:
            data = path.read_bytes()
        except OSError as exc:
            doc.findings.append(Finding(
                "parse_failed", f"无法读取 BPMN 文件：{exc}", {}, severity="warn"))
            return doc

        if len(data) > _MAX_XML_BYTES:
            doc.findings.append(Finding(
                "file_too_large",
                f"BPMN XML 大于 {_MAX_XML_BYTES // 1024 // 1024} MiB，已拒绝解析",
                {"kind": "xml", "pointer": "/"}, severity="warn"))
            return doc

        # ElementTree 不会主动取外部 DTD，但内部实体仍可能放大。BPMN 2.0 实例不需要
        # DTD，直接拒绝比尝试区分合法/恶意声明更稳妥。
        upper = data[: min(len(data), 64 * 1024)].upper()
        if b"<!DOCTYPE" in upper or b"<!ENTITY" in upper:
            doc.findings.append(Finding(
                "unsafe_xml", "BPMN 含 DTD/ENTITY 声明，已为避免实体展开攻击而拒绝解析",
                {"kind": "xml", "pointer": "/"}, severity="warn"))
            return doc

        try:
            root = ET.fromstring(data)
        except ET.ParseError as exc:
            locator: dict[str, Any] = {"kind": "xml", "pointer": "/"}
            if getattr(exc, "position", None):
                locator["line"], locator["column"] = exc.position
            doc.findings.append(Finding(
                "parse_failed", f"BPMN XML 语法错误：{exc}", locator, severity="warn"))
            return doc

        root_name = _local(root.tag)
        namespace = _namespace(root.tag)
        doc.meta.update({
            "namespace": namespace,
            "definitions_id": root.attrib.get("id", ""),
            "target_namespace": root.attrib.get("targetNamespace", ""),
        })
        if root_name != "definitions":
            doc.findings.append(Finding(
                "unexpected_root", f"BPMN 根节点应为 definitions，实际为 {root_name!r}",
                {"kind": "xml", "pointer": f"/{root_name}"}, severity="warn"))

        processes: list[dict[str, Any]] = []
        all_nodes: list[dict[str, Any]] = []
        all_flows: list[dict[str, Any]] = []
        order = 0
        process_elements = [e for e in root.iter() if _local(e.tag) == "process"]
        for process_no, process in enumerate(process_elements, start=1):
            process_id = process.attrib.get("id") or f"process-{process_no}"
            process_name = process.attrib.get("name") or process_id
            if "id" not in process.attrib:
                doc.findings.append(Finding(
                    "missing_id", f"第 {process_no} 个 process 没有 id，临时使用 {process_id}",
                    _locator(process_id, "process", process_id), severity="warn"))

            lanes = _lane_members(process)
            nodes: list[dict[str, Any]] = []
            flows: list[dict[str, Any]] = []
            anonymous = 0
            for element in process.iter():
                element_type = _local(element.tag)
                category = _category(element_type)
                if category is None:
                    continue
                element_id = element.attrib.get("id")
                if not element_id:
                    anonymous += 1
                    element_id = f"anonymous-{category}-{anonymous}"
                    doc.findings.append(Finding(
                        "missing_id",
                        f"流程 {process_id} 的 {element_type} 没有 id，临时使用 {element_id}",
                        _locator(process_id, element_type, element_id), severity="warn"))
                node = {
                    "id": element_id,
                    "name": element.attrib.get("name") or element_id,
                    "type": element_type,
                    "category": category,
                    "process_id": process_id,
                    "incoming": _child_texts(element, "incoming"),
                    "outgoing": _child_texts(element, "outgoing"),
                    "documentation": _documentation(element),
                    "event_definitions": _event_definitions(element),
                    "lanes": lanes.get(element_id, []),
                    "attributes": {
                        _local(k): v for k, v in element.attrib.items()
                        if _local(k) not in {"id", "name"}
                    },
                }
                nodes.append(node)

            node_by_id = {node["id"]: node for node in nodes}
            for element in process.iter():
                if _local(element.tag) != "sequenceFlow":
                    continue
                flow_id = element.attrib.get("id") or f"flow-{len(flows) + 1}"
                if "id" not in element.attrib:
                    doc.findings.append(Finding(
                        "missing_id",
                        f"流程 {process_id} 的 sequenceFlow 没有 id，临时使用 {flow_id}",
                        _locator(process_id, "sequenceFlow", flow_id), severity="warn"))
                source = element.attrib.get("sourceRef", "")
                target = element.attrib.get("targetRef", "")
                condition = _child_text(element, "conditionExpression")
                flow = {
                    "id": flow_id,
                    "name": element.attrib.get("name") or flow_id,
                    "type": "sequenceFlow",
                    "process_id": process_id,
                    "sourceRef": source,
                    "targetRef": target,
                    "condition": condition,
                    "documentation": _documentation(element),
                    "attributes": {
                        _local(k): v for k, v in element.attrib.items()
                        if _local(k) not in {"id", "name", "sourceRef", "targetRef"}
                    },
                }
                flows.append(flow)
                for ref_name, ref in (("sourceRef", source), ("targetRef", target)):
                    if not ref or ref not in node_by_id:
                        doc.findings.append(Finding(
                            "dangling_reference",
                            f"sequenceFlow {flow_id} 的 {ref_name}={ref!r} 未指向已解析节点",
                            _locator(process_id, "sequenceFlow", flow_id), severity="warn"))

            process_data = {
                "id": process_id,
                "name": process_name,
                "isExecutable": _bool_attr(process.attrib.get("isExecutable")),
                "documentation": _documentation(process),
                "nodes": nodes,
                "sequenceFlows": flows,
            }
            processes.append(process_data)
            all_nodes.extend(nodes)
            all_flows.extend(flows)

            process_render = (
                f"BPMN 流程 {process_name}（id={process_id}）："
                f"{len(nodes)} 个节点，{len(flows)} 条 sequenceFlow"
            )
            if process_data["documentation"]:
                process_render += f"。{process_data['documentation']}"
            doc.chunks.append(make_chunk(
                doc_id=f"bpmn:{process_id}", file_id=file_id, file_name=path.name,
                locator=_locator(process_id, "process", process_id),
                render=process_render, raw=process_data, order=order,
                tags=["bpmn", "process"]))
            order += 1

            for node in nodes:
                lane_text = "；泳道=" + "、".join(
                    lane["name"] for lane in node["lanes"]) if node["lanes"] else ""
                relation_text = ""
                if node["incoming"]:
                    relation_text += "；incoming=" + "、".join(node["incoming"])
                if node["outgoing"]:
                    relation_text += "；outgoing=" + "、".join(node["outgoing"])
                doc_text = f"；说明={node['documentation']}" if node["documentation"] else ""
                definition_text = (
                    "；事件定义="
                    + "、".join(item["type"] for item in node["event_definitions"])
                    if node["event_definitions"] else ""
                )
                doc.chunks.append(make_chunk(
                    doc_id=f"bpmn:{process_id}:{node['id']}", file_id=file_id,
                    file_name=path.name,
                    locator=_locator(process_id, node["type"], node["id"]),
                    render=(f"BPMN {node['category']} {node['name']}"
                            f"（{node['type']}，id={node['id']}）"
                            f"；流程={process_name}{lane_text}{relation_text}"
                            f"{definition_text}{doc_text}"),
                    raw=node, order=order, tags=["bpmn", node["category"], node["type"]]))
                order += 1

            for flow in flows:
                source_name = node_by_id.get(flow["sourceRef"], {}).get("name", flow["sourceRef"])
                target_name = node_by_id.get(flow["targetRef"], {}).get("name", flow["targetRef"])
                condition_text = f"；条件={flow['condition']}" if flow["condition"] else ""
                label_text = (f"；名称={flow['name']}"
                              if flow["name"] != flow["id"] else "")
                doc.chunks.append(make_chunk(
                    doc_id=f"bpmn:{process_id}:{flow['id']}", file_id=file_id,
                    file_name=path.name,
                    locator=_locator(process_id, "sequenceFlow", flow["id"]),
                    render=(f"BPMN sequenceFlow {flow['id']}：{source_name}"
                            f"（{flow['sourceRef']}） → {target_name}（{flow['targetRef']}）"
                            f"{label_text}{condition_text}"),
                    raw=flow, order=order,
                    tags=["bpmn", "sequenceFlow", "relation"]
                         + (["rule"] if flow["condition"] else [])))
                order += 1

        doc.structured = {
            "processes": processes,
            "nodes": all_nodes,
            "sequenceFlows": all_flows,
        }
        if not processes:
            doc.findings.append(Finding(
                "no_processes", "BPMN 文件里没有 process", {"kind": "xml", "pointer": "/"},
                severity="warn"))
        return doc


def _local(tag: str) -> str:
    return tag.rsplit("}", 1)[-1].rsplit(":", 1)[-1]


def _namespace(tag: str) -> str:
    return tag[1:].split("}", 1)[0] if tag.startswith("{") and "}" in tag else ""


def _category(element_type: str) -> str | None:
    if element_type in _TASK_TYPES or element_type.endswith("Task"):
        return "task"
    if element_type in _EVENT_TYPES or element_type.endswith("Event"):
        return "event"
    if element_type in _GATEWAY_TYPES or element_type.endswith("Gateway"):
        return "gateway"
    return None


def _child_text(element: ET.Element, child_name: str) -> str:
    for child in element:
        if _local(child.tag) == child_name:
            return " ".join("".join(child.itertext()).split())
    return ""


def _child_texts(element: ET.Element, child_name: str) -> list[str]:
    return [text for child in element if _local(child.tag) == child_name
            and (text := " ".join("".join(child.itertext()).split()))]


def _documentation(element: ET.Element) -> str:
    return _child_text(element, "documentation")


def _event_definitions(element: ET.Element) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for child in element:
        child_type = _local(child.tag)
        if not child_type.endswith("EventDefinition"):
            continue
        out.append({
            "type": child_type,
            "attributes": {_local(key): value for key, value in child.attrib.items()},
            "text": " ".join("".join(child.itertext()).split()),
        })
    return out


def _lane_members(process: ET.Element) -> dict[str, list[dict[str, str]]]:
    out: dict[str, list[dict[str, str]]] = {}
    for lane in process.iter():
        if _local(lane.tag) != "lane":
            continue
        lane_id = lane.attrib.get("id", "")
        lane_name = lane.attrib.get("name") or lane_id
        for child in lane:
            if _local(child.tag) != "flowNodeRef":
                continue
            ref = " ".join("".join(child.itertext()).split())
            if ref:
                out.setdefault(ref, []).append({"id": lane_id, "name": lane_name})
    return out


def _locator(process_id: str, element_type: str, element_id: str) -> dict[str, Any]:
    pointer = f'/definitions/process[@id="{process_id}"]'
    if element_type != "process":
        pointer += f'//{element_type}[@id="{element_id}"]'
    return {
        "kind": "xml",
        "pointer": pointer,
        "process": process_id,
        "element": element_type,
        "id": element_id,
    }


def _bool_attr(value: str | None) -> bool | None:
    if value is None:
        return None
    return value.strip().lower() in {"true", "1"}
